src/plugins/cpptemplate/cpptemplate.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file
+ *
+ * @brief Source for cpptemplate plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include "cpptemplate.hpp"
+#include "cpptemplate_delegate.hpp"
+
+#include <kdbhelper.h>
+
+using elektra::CppTemplateDelegate;
+
+using CppKey = kdb::Key;
+using CppKeySet = kdb::KeySet;
+
+namespace
+{
+
+/**
+ * @brief This function returns a key set containing the plugin contract.
+ *
+ * @return A key set specifying the capabilities of the plugin
+ */
+CppKeySet getContract ()
+{
+	return CppKeySet{ 30,
+			  keyNew ("system/elektra/modules/cpptemplate", KEY_VALUE, "cpptemplate plugin waits for your orders", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/open", KEY_FUNC, elektraCppTemplateOpen, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/close", KEY_FUNC, elektraCppTemplateClose, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/get", KEY_FUNC, elektraCppTemplateGet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/set", KEY_FUNC, elektraCppTemplateSet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/error", KEY_FUNC, elektraCppTemplateError, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/checkconf", KEY_FUNC, elektraCppTemplateCheckConfig, KEY_END),
+#include ELEKTRA_README (cpptemplate)
+			  keyNew ("system/elektra/modules/cpptemplate/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END),
+			  KS_END };
+}
+
+} // end namespace
+
+extern "C" {
+
+typedef Delegator<CppTemplateDelegate> delegator;
+
+/** @see elektraDocOpen */
+int elektraCppTemplateOpen (Plugin * handle, Key * key)
+{
+	// After the call to `delegator::open` you can retrieve a pointer to the delegate via `coderDelegator::get (handle)`
+	return delegator::open (handle, key);
+}
+
+/** @see elektraDocClose */
+int elektraCppTemplateClose (Plugin * handle, Key * key)
+{
+	return delegator::close (handle, key);
```

Comment:
pass CppKey here (and also in Open)?

Reply:
As far as I can tell that would only work if we change the definition of `open` in `kdbplugin.hpp`? Should I really do that or do? Wouldn’t that break the current C++ plugin API?

Reply:
You are right, the change does not make much sense if this is hidden in a macro anyway. For user code, however, it would be nice if devs are only confronted with CppKey(Set). In particular code as in src/plugins/type/type.cpp line 95 or src/plugins/ccode/ccode.cpp lines 71-75, 85-88 should be avoided.

Such a change breaks the API but it does not break the ABI because it is header-only. So people need to adapt their code but already compiled code still works.

Reply:
Oh, I see. In that case I **might** change the interface in a separate pull request.